KM GVS documentation (#7903)
broadinstitute/gatk: scripts/variantstore/gvs-overview.md

# Genomic Variant Store Pipeline
 
| Pipeline Version | Date Updated | Documentation Authors | Questions or Feedback |
| :----: | :---: | :----: | :--------------: |

> | [GvsJointVariantCalling](https://github.com/broadinstitute/gatk/blob/rc-vs-483-beta-user-wdl/scripts/variantstore/wdl/GvsJointVariantCalling.wdl) | June, 2022 | [Kaylee Mathews](mailto:kmathews@broadinstitute.org) and [Aurora Cremer](mailto:aurora@broadinstitute.org) | If you have questions or feedback, contact the [Broad Variants team](mailto:variants@broadinstitute.org) |

Author: Will update all the links to the wrapper WDL once the [PR](https://github.com/broadinstitute/gatk/pull/7894) is merged

> | [GvsJointVariantCalling](https://github.com/broadinstitute/gatk/blob/rc-vs-483-beta-user-wdl/scripts/variantstore/wdl/GvsJointVariantCalling.wdl) | June, 2022 | [Kaylee Mathews](mailto:kmathews@broadinstitute.org) and [Aurora Cremer](mailto:aurora@broadinstitute.org) | If you have questions or feedback, contact the [Broad Variants team](mailto:variants@broadinstitute.org) |

Reviewer: change email to warp-pipelines-help@broadinstitute.org

Author: @kbergin suggested we use the Variants email. Is that still what you'd like to do Kylee or would you rather use the WARP help email?

> ![Diagram depicting the Broad Genomic Variant Store workflow. Sample GVCF files are imported into the core data model. A filtering model is trained using Variant Quality Score Recalibration, or VQSR, and then used to extract cohorts and produce sharded joint VCF files. Each step integrates BigQuery and GATK tools.](/scripts/variantstore/genomic-variant-store_diagram.png)

Author: This is a placeholder figure for now until we have time to create one with a little more detail.

## Introduction to the Genomic Variant Store pipeline

> The [Genomic Variant Store (GVS)](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/gvs-product-sheet.pdf) was developed by the Data Sciences Platform at the Broad Institute of MIT and Harvard as a solution for variant discovery at scales previously inaccessible to most researchers due to large cohort sizes and high computational and financial costs. The Broad GVS is powered by BigQuery and currently supports the creation of joint callsets of up to 100,000 human genomes faster, cheaper, and more reliably than existing solutions.

Reviewer: ```suggestion
The [Genomic Variant Store (GVS)](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/gvs-product-sheet.pdf) was developed by the Data Sciences Platform at the Broad Institute of MIT and Harvard as a solution for variant discovery at scales previously inaccessible to most researchers due to large cohort sizes and high computational and financial costs. The Broad GVS is powered by BigQuery and creates joint callsets of up to 100,000 human genomes faster, more cheaply, and more reliably than existing solutions.
```

Reviewer: I think you want cheaply here to match the verb

Reviewer 2: The first sentence seems complex and wordy to me, maybe it's the use of "inaccessible".   You're the expert here, so whatever you think is best.  But maybe remove the attribution and simplify the motivation?  

> we developed to address scale, cost and operational challenges in joint calling in large whole genome joint callsets.

Reviewer 2: Also 

> than existing solutions

If we're going to say this, we might need to show our work (e.g. compare to GLnexus, sentieon, etc).  Maybe it would be better to say "previous solutions" or something?

Author: Great suggestions. I made the suggested changes and simplified the first sentence.

The [GVS pipeline](https://github.com/broadinstitute/gatk/blob/rc-vs-483-beta-user-wdl/scripts/variantstore/wdl/GvsJointVariantCalling.wdl) is an open-source, cloud-optimized workflow for joint calling at a large scale using the Broad GVS. The workflow takes in up to 100,000 single sample gVCF files and combines them into a variant filtering model driven by machine learning. The model is applied to the data, and a sharded joint VCF with variant calls is output.

> The filtering model is based on the [WARP Joint Genotyping workflow](https://github.com/broadinstitute/warp/blob/master/pipelines/broad/dna_seq/germline/joint_genotyping/JointGenotyping.wdl) and uses the [Variant Quality Score Recalibration (VQSR)](https://gatk.broadinstitute.org/hc/en-us/articles/360035531612) technique which uses machine learning to model the technical profile of variants to flag probable artifacts.

Reviewer: Maybe add a little more summary description about what the pipeline does? Does it automatically load your input data to the BigQuery dataset you create and then perform the variant filtering model?

Author: Removed several mentions of up to 100k and expanded the intro section.

## Quickstart table

The following table provides a quick overview of the GVS pipeline features:

| Pipeline features | Description | Source | 
| --- | --- | --- |

> | Overall workflow | Joint Calling with the Broad Genomic Variant Store using a machine learning-powered filter model | Code available from [GitHub](https://github.com/broadinstitute/gatk/blob/rc-vs-483-beta-user-wdl/scripts/variantstore/wdl/GvsJointVariantCalling.wdl) |

Reviewer: ```suggestion
| Overall workflow | Joint calling with the Broad Genomic Variant Store using a machine learning-powered filter model | Code available from [GitHub](https://github.com/broadinstitute/gatk/blob/rc-vs-483-beta-user-wdl/scripts/variantstore/wdl/GvsJointVariantCalling.wdl) |
```

Reviewer: Does the model have a specific name? If so, I think I'd be curious what it is if I was glancing at the table.

| Workflow language | WDL 1.0 | [openWDL](https://github.com/openwdl/wdl) |
| Genomic reference sequence | GRCh38 (hg38) human genome primary sequence | Genome Reference Consortium [GRCh38](https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.39) |
| Data input file format | File format in which input data is provided | [gVCF](https://gatk.broadinstitute.org/hc/en-us/articles/360035531812) |
| Data output file formats | File formats in which outputs are provided | [VCF](https://gatk.broadinstitute.org/hc/en-us/articles/360035531692) and associated index files; TXT (manifest) |

## Setup

### Getting started with Terra

> The GVS workflow is available on Terra, a cloud-based platform for biomedical research. After registering, you can access the [GVS workspace](LINK), which contains instructions for performing joint calling with the Broad GVS using example data uploaded to the workspace. For more information on using the Terra platform, see the [Terra Support Center](https://support.terra.bio/hc/en-us).

Author: Changed all instances of "Broad GVS" to "GVS"

> If you are new to Terra, you’ll need to [register for a Terra account](https://support.terra.bio/hc/en-us/articles/360028235911). Beta testers of the GVS will be given access to a Broad-managed billing project, but to use Terra for other projects, you’ll need to [set up a billing project](https://support.terra.bio/hc/en-us/articles/360046295092).

Reviewer 2: Are we going to give them access to a Broad billing project?  That seems… risky since they could run up a massive bill in there doing other work, making mistakes, etc.  Can we use the free-credits approach here?

Reviewer 3: This is going to be a Kylee question!

### Workflow requirements

#### Set up BigQuery

To start using the GVS workflow in Terra, you need to create a BigQuery dataset with permissions that will allow Terra to access the dataset by following the instructions below:

**1.** First, **create a dataset in BigQuery** by following the instructions in [Creating datasets](https://cloud.google.com/bigquery/docs/datasets) from the BigQuery documentation.

**2.** Find your Terra proxy group by **selecting the main menu** (three horizontal line icon) at the top left of any Terra page. Next, **click on your account name**, followed by **Profile**. Here, you should see a field called **Proxy Group**, which lists your proxy group underneath. For more information on proxy groups in Terra, see [Pet service accounts and proxy groups](https://support.terra.bio/hc/en-us/articles/360031023592). 

**3.** Grant your proxy group the **BigQuery Data Editor** role on the dataset you created in Step 1 by following the instructions in [Controlling access to datasets](https://cloud.google.com/bigquery/docs/dataset-access-controls) from the BigQuery documentation.

> **4.** Grant your proxy group the **BigQuery Data Editor**, **BigQuery Job User**, and **BigQuery Read Session User** roles on the Google Project that contains the dataset you created in Step 1 by following the instructions in [Manage access to projects, folders, and organizations](https://cloud.google.com/iam/docs/granting-changing-revoking-access) from the Google Cloud documentation.

Reviewer: Is there anything they should see after their done following these steps? Anything that would tell them that they did it right?

Author: I added a bit of text to the Quickstart document to describe how users can see if they've setup the permissions correctly.

#### Reblocked gVCF files

> The GVS workflow takes in reblocked gVCF files as input. If your files are not already reblocked, you can reblock them using the [WARP reblocking workflow](https://github.com/broadinstitute/warp/blob/master/pipelines/broad/dna_seq/germline/joint_genotyping/reblocking/ReblockGVCF.wdl), which is configured in the [ReblockGVCF Terra workspace](https://app.terra.bio/#workspaces/warp-pipelines/ReblockGVCF). For more information about reblocking, check out [WARP Whole Genome and Exome Pipelines Produce Reblocked GVCFs](https://broadinstitute.github.io/warp/blog/tags/reblock/).

Reviewer: So does the input to the BigQuery need to be reblocked?

Author: I think we'll definitely make it clear on the diagram when we make a more detailed one that matches the style of other GATK stuff.

#### gVCF annotations

> 

Reviewer: Table descriptions might need a period at the end.

> Input gVCF files for the GVS workflow must include the annotations described in the table below:

Reviewer 2: Maybe it would be helpful to call out earlier that the gVCFs must have been created with a GATK "functional equivalence" compatible solution (basically, HaplotypeCaller, DRAGEN, Sention).  That's where these annotations come from and might make this easier to understand.    The methods folks might have good links for this

| Annotation | Description | Notes |
| --- | --- | --- |
| Ref | Reference allele | --- |
| Alt | Alternate allele | --- |
| AS_RAW_MQ, RAW_MQandDP, or RAW_MQ | RMS mapping quality (‘AS’: allele-specific) | Required for VQSR Data |
| AS_RAW_MQRankSum or Map_QUAL_RANK_SUM_KEY | Z-score from Wilcoxon rank sum test of alternate versus reference read mapping qualities | Required for VQSR Data |
| QUALapprox | Sum of PL[0] values; used to approximate the QUAL score | Required for VQSR Data |
| AS_QUALapprox | Allele-specific sum of PL[0] values; used to approximate the QUAL score | Required for VQSR Data |
| AS_SB_TABLE or STRAND_BIAS_BY_SAMPLE | Allele-specific forward/reverse read counts for strand bias tests | Required for VQSR Data |
| AS_VarDP, VarDP, or DP | Depth over variant genotypes, or read depth  (‘AS’: allele-specific) | Required for VQSR Data |
| call_GT | Genotype | --- |
| call_GQ | Genotype quality | --- |

### Inputs

> The GVS workflow inputs are specified in the Terra GVS beta workspace [workflow configuration](LINK).

Reviewer: ```suggestion
The GVS workflow inputs are described in the table below and are specified in the Terra GVS beta workspace [workflow configuration](LINK).
```

#### Sample data input

> The GVS workflow takes in reblocked single sample gVCF files and their corresponding index files as `input_vcfs` and `input_vcf_indexes`, respectively. While the GVS workflow takes in up to 100,000 single sample gVCF files as input, only datasets of up to 10,000 files are being used for beta testing.

Reviewer 2: Two things

1. as `input_vcfs` -- I think it's implied here that we're talking about the terra data model?  Maybe make that more clear esp if the audience is new to Terra

2. Again I would remove the 100k piece and just say "For the beta, cohort sizes of up to 10k samples are supported"

The [Terra GVS beta workspace](LINK) has been configured with example reblocked gVCF files that you can use to test the workflow. To test the workflow on your own sample data, you’ll need to create an additional data table in your workspace that specifies the locations of the files in the cloud. For step-by-step instructions on creating data tables in Terra, see [How to make a data table from scratch or a template](https://support.terra.bio/hc/en-us/articles/6197368140955). If your data is stored locally, you can upload it to your workspace bucket using the Data Uploader in Terra by following the instructions in the article, [How to use the Data Uploader](https://support.terra.bio/hc/en-us/articles/4419428208411). You’ll also need to ensure that you have a BigQuery dataset and Google project with the correct permissions and that your gVCF files are reblocked, as described in the [Workflow requirements](#workflow-requirements) section above.

#### Input descriptions

The table below describes the GVS workflow input variables:

| Input variable name | Description | Type |
| --- | --- | --- |

> | dataset_name | String describing the name of the BigQuery dataset. | String |

Reviewer: Will the data files specified in the input be added to this dataset? I'm new to use BigQuery as part of a workflow, so I'm just curious for more details on how this BigQuery dataset is used relative to the sample inputs.

Reviewer 3: a BQ dataset is a collection of tables. The data from the gvcf files will be added into specific (vet and ref_ranges) tables in the dataset.

| project_id | String describing the name of the Google project that contains the sample data. | String |

> | external_sample_names | Strings describing the unique sample IDs in the Terra data model. | Array of strings |

Reviewer: Are the samples we're describing here the ones from which the individual GVCFs were derived?

Reviewer 2: Wordy -- drop the "Strings describing…" e.g. 
 
> name of the Google project that contains the bigquery dataset.

Reviewer 3: re samples: 
This string will be exactly what shows up in the final joint VCF so however the user wants to see the data there. 

The external_sample_name is just meant to be the users identifier for the sample because in GVS we assign an id for efficiency.

Author: Added text to clarify what the `external_sample_names` variable is and removed the redundancies from the descriptions.

> | input_vcfs | Cloud paths to the sample gVCF files. | Array of files |

Reviewer 2: These two might be confusing to a new user… What we really want here is the column name from the data table that points to the reblocked gVCFs… (which is an array of files) right?

| input_vcf_indexes | Cloud paths to the sample gVCF index files. | Array of files |

> | callset_identifier | String used as the name of the filter model and as the prefix to the names of the BigQuery extract tables and final joint VCF shards. | String |

Author: Done

## Tasks and tools

The [GVS workflow](https://github.com/broadinstitute/gatk/blob/rc-vs-483-beta-user-wdl/scripts/variantstore/wdl/GvsJointVariantCalling.wdl) imports additional WDL scripts that contain the workflow tasks. Links to these additional WDL scripts (subtasks/subworkflows) are provided below.

Overall, the GVS workflow:

> 1. Imports sample genomes.

Reviewer 2: genomes -> gVCFs

1. Trains the filtering model.

> 1. Extracts VCF files.

Reviewer 2: Extracts joint call VCF files with filtering applied

To see specific tool parameters used in each step of the workflow, select the WDL link in the table; then find the task and view the `command {}` section of the task in the WDL script.

To view or use the exact tool software, see the task's Docker image, which is specified in the task `runtime {}` section as `docker: `.

> ### 1. Import sample genomes

Reviewer: If I understand this section correctly, it seems like the workflow is importing the GVCFs (variant info) rather than a genome? Is genome the preferred wording?

Reviewer 2: +1 I think this should be gVCFs

To import sample genomes, the GVS workflow calls two subworkflows detailed below.

> #### A. GvsAssignIds

Author: We want to add more information (i.e., individual tasks being called) for these subworkflows at some point, but we focused on getting this up with the info beta users need to run the workflow in Terra first.

> #### A. GvsAssignIds

Reviewer: You could consider making the headings a high level description of what the task is doing instead of the task name.

Author: Revamped this section and these headers are no longer in the doc.

> 

Reviewer: This might be silly, but could you explain a little more what an access pattern is? It comes up later on too when the workflow makes a new one. What exactly is it?

Reviewer 3: I think KC is right on that we should skip it---for people comfortable with moving data around, it reads as a way to move it more efficiently---I dont think it's a silly question at all; it's just more of a reason to skip mentioning it

> The [GvsAssignIds subworkflow (alias = AssignIds)](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/wdl/GvsAssignIds.wdl) assigns GVS-specific IDs to each of the samples in the sample set which are used as an access pattern. The IDs enable easy access to specific sample data. GvsAssignIds imports an additional workflow, [GvsCreateTables](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/wdl/GvsCreateTables.wdl).

Reviewer: Which input in the table above does this step take in?

Reviewer: Might be useful to say something like it takes in the BigQuery dataset name, sample names and project ID info to assign GVS-specific IDs...

Author: I simplified this section significantly. Let me know what you all think!

#### B. GvsImportGenomes

> The [GvsImportGenomes subworkflow (alias = ImportGenomes)](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/wdl/GvsImportGenomes.wdl) validates that sample gVCF files contain required annotations and loads samples into several tables in BigQuery. GvsImportGenomes imports an additional workflow, [GvsUtils (alias = Utils)](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/wdl/GvsUtils.wdl).

Reviewer 2: It doesn't feel like describing this Utils is helpful to the user, and we're not providing any additional info here.  I'd remove it

Author: I totally revamped this section. Now it's more of a high-level description of what the three major steps of the workflow do and what GATK tools are used to do it.

### 2. Train the filtering model

To train the filtering model, the GVS workflow calls two subworkflows detailed below.

> #### A. GvsCreateAltAllele

Reviewer 2: I think conceptually I'd put this under the "Loading Data" section, esp as we move into the "search" space, creating AA won't just be part of the model training.

Reviewer 3: Do we still make an alt allele table for use cases where we skip filter creation?

I think at least until we make the AA table incremental (and not just something we recreate each time for a new filter model) it fits fine under the filter umbrella

Author: I think I've updated this section in such a way that it doesn't matter for now?

> The [GvsCreateAltAllele subworkflow (alias = CreateAltAllele)](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/wdl/GvsCreateAltAllele.wdl) splits alternate alleles and calculates additional annotations to be used for filtering. GvsCreateAltAllele imports an additional workflow, [GvsUtils (alias = Utils)](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/wdl/GvsUtils.wdl).

Reviewer: It might be useful to list subworkflows separate from the description paragraph. The way it reads now, I almost expect the sentence about the additional workflow to be followed up with a description of what that subworkflow is doing.

We've done this for some of the other GATK pipelines like the WGS. We could use a similar technique here or try to a new way to break it up.

Author: Totally revamped the tasks and tools section of this doc. Let me know what you think!

#### B. GvsCreateFilterSet

> 

Reviewer: What is the output of this task? The next section talks about VCFs, but we've been working with GVCF up till this point.

Reviewer 3: another table in the dataset -- not something the user needs to understand, but it's the output of this that we use to create the filter

Reviewer 3: this is the first part of the BQ only part of the workflow. We are taking data from one table (or several, if the number of samples is >4k) and moving it into another table (partially for data access patterns!!!)

Author: I simplified this section significantly so I don't think your question is answered in the doc @ekiernan. However, I hope by simplifying it, it makes it seem less like something users need to know the inner workings of?

> The [GvsCreateFilterSet subworkflow (alias = CreateFilterSet)](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/wdl/GvsCreateFilterSet.wdl) creates a non-deterministic filtering model using [VQSR](https://gatk.broadinstitute.org/hc/en-us/articles/360035531612) and annotations from a random subset of the input samples. SNPs are recalibrated using the annotations `AS_FS`, `AS_ReadPosRankSum`, `AS_MQRankSum`, `AS_QD`, and `AS_SOR`. Indels are recalibrated using the annotations `AS_FS`, `AS_ReadPosRankSum`, `AS_MQRankSum`, `AS_QD`, `AS_SOR`, and `AS_MQ`. GvsCreateFilterSet imports two additional WDL scripts, [GvsWarpTasks (alias = Tasks)](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/wdl/GvsWarpTasks.wdl) and [GvsUtils (alias = Utils)](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/wdl/GvsUtils.wdl).

Reviewer 2: I wouldn't put "non-deterministic" in there -- it may be true in certain circumstances that are very nuanced… but it's not a feature or somethign we want to highlight.  If anything we could pull out something from the VQSR docs (ie a "variational bayes" or something)

Author: Removed

### 3. Extract VCF files

To extract joint VCF files, the GVS workflow calls two subworkflows detailed below.

#### A. GvsPrepareRangesCallset

> The [GvsPrepareRangesCallset subworkflow (alias = PrepareRangesCallset)](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/wdl/GvsPrepareRangesCallset.wdl) transposes data into a new access pattern based on genomic coordinates.

Reviewer 2: > transposes data into a new access pattern based on genomic coordinates.

I'm not sure what this would mean to beta users.  Maybe just better to say what it does at a high level like "subsets data into temporary tables for efficient extract" or something
